dotnet/src/Extensions/PromptTemplate.Handlebars/Helpers/KernelHelpers/KernelFunctionHelpers.cs

```diff
@@ -0,0 +1,210 @@
+                KernelFunction function = kernel.Plugins.GetFunction(functionMetadata.PluginName, functionMetadata.Name);
+
+                // Invoke the function and write the result to the template
+                return InvokeKernelFunction(kernel, function, executionContext, cancellationToken);
+            });
+    }
+
+    /// <summary>
+    /// Checks if handlebars argument is a valid type for the function parameter.
+    /// Must satisfy one of the following:
+    /// Types are an exact match.
+    /// Argument is any kind of numeric type if function parameter requires a numeric type.
+    /// Argument type is an object (this covers complex types).
+    /// Function parameter is a generic type.
+    /// </summary>
+    /// <param name="parameterType">Function parameter type.</param>
+    /// <param name="argument">Handlebar argument.</param>
+    private static bool IsExpectedParameterType(KernelParameterMetadata parameterType, object argument)
+    {
+        var actualParameterType = parameterType.ParameterType.TryGetGenericNullableType(out var nullableType) ? nullableType : parameterType.ParameterType;
+
+        bool parameterIsNumeric = KernelHelpersUtils.IsNumericType(actualParameterType)
+            || (parameterType.Schema?.RootElement.TryGetProperty("type", out JsonElement typeProperty) == true && typeProperty.GetString() == "number");
+
+        bool argIsNumeric = KernelHelpersUtils.IsNumericType(argument?.GetType())
+            || KernelHelpersUtils.TryParseAnyNumber(argument?.ToString());
+
+        return actualParameterType is null
+            || actualParameterType == argument?.GetType()
+            || (parameterIsNumeric && argIsNumeric);
+    }
+
+    /// <summary>
+    /// Processes the hash arguments passed to a Handlebars helper function.
+    /// </summary>
+    /// <param name="functionMetadata">Metadata for the function being invoked.</param>
+    /// <param name="executionContext">Arguments maintained in the executing context.</param>
+    /// <param name="handlebarsArguments">Arguments passed to the Handlebars helper.</param>
+    /// <param name="nameDelimiter">The character used to delimit the plugin name and function name in a Handlebars template.</param>
+    /// <exception cref="KernelException">Thrown when a required parameter is missing.</exception>
+    private static void ProcessHashArguments(
+        KernelFunctionMetadata functionMetadata,
+        KernelArguments executionContext,
+        IDictionary<string, object>? handlebarsArguments,
+        string nameDelimiter)
+    {
+        // Prepare the input parameters for the function
+        foreach (var param in functionMetadata.Parameters)
+        {
+            var fullyQualifiedParamName = functionMetadata.Name + nameDelimiter + param.Name;
+            var value = handlebarsArguments is not null && (handlebarsArguments.TryGetValue(param.Name, out var val) || handlebarsArguments.TryGetValue(fullyQualifiedParamName, out val)) ? val : null;
```

Comment:
Shouldn't we first use a specific parameter like "fullyQualifiedParamName" to resolve the argument, and if not found, use a less specific one "param.Name"?